### PR TITLE
Add Servant.Client.Free

### PR DESCRIPTION
**Pull request by phadej**
this is useful in testing, it nothing else.

### CODE REVIEW

File: servant-client-core/src/Servant/Client/Core/Internal/RunClient.hs [hunk 1]
```diff
@@ -48,3 +51,12 @@ decodedAs response contentType = do
     Right val -> return val
   where
     accept = toList $ contentTypes contentType
+
+instance ClientF ~ f => RunClient (Free f) where
```
Comment (phadej) [inline servant-client-core/src/Servant/Client/Core/Internal/RunClient.hs]: dunno if should rather have `RunClientF f => RunClient (Free f)`

File: servant-client-core/src/Servant/Client/Core/Internal/RunClient.hs [hunk 2]
```diff
@@ -48,3 +51,12 @@ decodedAs response contentType = do
...
+    runRequest req  = liftF (RunRequest req id)
+    streamingRequest req = liftF (StreamingRequest req id)
+    throwServantError = liftF . Throw
+    catchServantError x h = go x where
```
Comment (phadej) [inline servant-client-core/src/Servant/Client/Core/Internal/RunClient.hs]: this is not particularly correct, as `RunRequest` could throw things catchable by `catchServantError` too.

**Comment by jkarni** [comment]
Is this missing a file (`ClientF`)?

**Comment by phadej** [comment]
oops

File: servant-client-core/src/Servant/Client/Core/Internal/ClientF.hs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{-# LANGUAGE DeriveFunctor #-}
+module Servant.Client.Core.Internal.ClientF where
+
+import Servant.Client.Core.Internal.Request
+
+data ClientF a
+    = RunRequest Request (Response -> a)
+    | StreamingRequest Request (StreamingResponse -> a)
+    | Throw ServantError
```
Comment (jkarni) [inline servant-client-core/src/Servant/Client/Core/Internal/ClientF.hs]: Maybe adding a `Catch` constructor would help fix the problem with `catchServantError` that you mention below?

Comment (phadej) [reply servant-client-core/src/Servant/Client/Core/Internal/ClientF.hs]: `catch` isn't an algebraic effect. So it won't really work out well.

**Comment by phadej** [comment]
@jkarni @alpmestan I'd rather remove `catchServantError`, we don't use it itself, and I doubt anyone else

**Comment by jkarni** [comment]
+1 to removing it. On my phone so can't review more carefully, but I think
that was my only concern.

**Comment by alpmestan** [comment]
LGTM, and I don't mind the removal of `catchServantError` as well.

**Pull request merged**